src/cli/compat_mode.c

```diff
@@ -0,0 +1,137 @@
+/**
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+#include <err.h>
+#include <libgen.h>
+#undef basename /* Use the GNU version of basename. */
+#include <stdlib.h>
+
+#include "cli.h"
+#include "compat_mode.h"
+
+static void filter_by_major_version(bool, const struct nvc_driver_info *, char * [], size_t *);
+static int get_compat_library_path(struct error *, const char * [], size_t, char **);
+
+int
+update_compat_libraries(struct nvc_context *ctx, struct nvc_container *cnt, const struct nvc_driver_info *info) {
+        if (cnt->flags & OPT_CUDA_COMPAT_MODE_DISABLED) {
+                return 0;
+        }
+        if (cnt->libs == NULL || cnt->nlibs == 0) {
+                return 0;
+        }
+        size_t nlibs = cnt->nlibs;
+        char **libs = array_copy(&ctx->err, (const char * const *)cnt->libs, cnt->nlibs);
+        if (libs == NULL) {
+                return -1;
+        }
+
+        // For cuda-compat-mode=mount, we also allow compat libraries with a LOWER major versions.
+        bool allow_lower_major_versions = (cnt-> flags & OPT_CUDA_COMPAT_MODE_MOUNT);
+        filter_by_major_version(allow_lower_major_versions, info, libs, &nlibs);
+
+        // We free the previously allocated libs to allow these to be reassigned if required.
+        free(cnt->libs);
+        cnt->libs = NULL;
+        cnt->nlibs = 0;
+
+        if (cnt->flags & OPT_CUDA_COMPAT_MODE_LDCONFIG) {
+                if (get_compat_library_path(&ctx->err, (const char **)libs, nlibs, &cnt->cuda_compat_dir) < 0) {
+                        goto fail;
+                }
+                // For cuda-compat-mode=ldconfig we don't require the library
+                // paths since we already have the compat dir stored.
+                free(libs);
+        }
+        if (cnt-> flags & OPT_CUDA_COMPAT_MODE_MOUNT) {
+                // Munting is handled later. We only need to update the required
+                // container information.
+                cnt->libs = libs;
+                cnt->nlibs = nlibs;
+        }
+        return (0);
+fail:
+        free(libs);
+        return (-1);
+}
```

Comment:
Is there any harm in not freeing the `cnt->libs` array in this function when `cuda-compat-mode=ldconfig`? If it is not required, then what about the below suggestion for readability sake? 
```suggestion
        filter_by_major_version(allow_lower_major_versions, info, libs, &nlibs);

        // Use filtered library list
        free(cnt->libs);
        cnt->libs = libs;
        cnt->nlibs = nlibs;

        if (cnt->flags & OPT_CUDA_COMPAT_MODE_LDCONFIG) {
                if (get_compat_library_path(&ctx->err, (const char **)libs, nlibs, &cnt->cuda_compat_dir) < 0) {
                        return (1);
                }
        }
        return (0);
}
```